pkg/agent/puller.go

```diff
@@ -1,73 +1,130 @@
+)
+
+// sentinel to represent model absence
+var remove v1.ModelSpec = v1.ModelSpec{}
+
 type Puller struct {
-	ChannelMap map[string]Channel
-	Downloader Downloader
+	channelMap  map[string]ModelChannel
+	completions chan string
+	Downloader  Downloader
 }
 
-type Channel struct {
-	EventChannel chan EventWrapper
+type ModelOp struct {
+	ModelName string
+	Op        OpType
+	Spec      *v1.ModelSpec
 }
 
-func (p *Puller) AddModel(modelName string) Channel {
-	// TODO: Figure out the appropriate buffer-size for this
-	// TODO: Check if event Channel exists
-	eventChannel := make(chan EventWrapper, 20)
-	channel := Channel{
-		EventChannel: eventChannel,
+func StartPuller(downloader Downloader, commands <-chan ModelOp) {
+	puller := Puller{
+		channelMap:  make(map[string]ModelChannel),
+		completions: make(chan string, 128),
+		Downloader:  downloader,
 	}
-	go p.modelProcessor(modelName, channel.EventChannel)
-	p.ChannelMap[modelName] = channel
-	return p.ChannelMap[modelName]
+	go puller.processCommands(commands)
 }
 
-func (p *Puller) RemoveModel(modelName string) error {
-	channel, ok := p.ChannelMap[modelName]
-	if ok {
-		close(channel.EventChannel)
-		delete(p.ChannelMap, modelName)
+func (p *Puller) processCommands(commands <-chan ModelOp) {
+	// channelMap accessed only by this goroutine
+	var finished bool
+	for {
+		select {
+		case modelOp, ok := <-commands:
+			if !ok {
+				finished = true
+				//commands = nil //TODO tbd
+			} else {
+				switch modelOp.Op {
+				case Add:
+					p.enqueueModelOp(modelOp.ModelName, modelOp.Spec)
+				case Remove:
+					p.enqueueModelOp(modelOp.ModelName, &remove)
+				}
+			}
+		case completed := <-p.completions:
+			p.modelOpComplete(completed, finished)
+		}
 	}
-	if err := storage.RemoveDir(filepath.Join(p.Downloader.ModelDir, modelName)); err != nil {
-		return fmt.Errorf("failing to delete model directory: %v", err)
+}
+
+type ModelChannel struct {
+	modelOps    chan *v1.ModelSpec
+	opsInFlight int
+}
+
+func (p *Puller) enqueueModelOp(modelName string, modelSpec *v1.ModelSpec) {
+	modelChan, ok := p.channelMap[modelName]
+	if !ok {
+		modelChan = ModelChannel{
+			modelOps: make(chan *v1.ModelSpec, 8),
+		}
+		go p.modelProcessor(modelName, modelChan.modelOps)
+		p.channelMap[modelName] = modelChan
+	}
+	modelChan.opsInFlight += 1
+	modelChan.modelOps <- modelSpec
+}
+
+func (p *Puller) modelOpComplete(modelName string, closed bool) {
+	modelChan, ok := p.channelMap[modelName]
+	if ok {
+		modelChan.opsInFlight -= 1
+		if modelChan.opsInFlight == 0 {
+			close(modelChan.modelOps)
+			delete(p.channelMap, modelName)
+			if closed && len(p.channelMap) == 0 {
+				// this was the final completion, close the channel
+				close(p.completions)
+			}
+		}
+	} else {
+		//TODO log warning, shouldn't happen
 	}
-	return nil
 }
 
-func (p *Puller) modelProcessor(modelName string, events chan EventWrapper) {
+func (p *Puller) modelProcessor(modelName string, ops <-chan *v1.ModelSpec) {
```

Comment:
The idea to make the channel able to receive and send events is so that I can handle retries (by writing back to the same channel), is there a strategy to handle retries? I guess writing back to same channel is a bad design, maybe?